### PR TITLE
Category in one place

### PR DESCRIPTION
**Pull request by escopecz**
This PR removes all the category menu items from left menu and adds one Category menu item to the right menu.

### How to test
1. Apply this PR and clear cache.
2. Make sure menu items are like described above.
3. You can see all categories from all bundles at one place when you open the Categories form the right menu.
4. You can filter by Type (by bundles)
5. You can create new categories and assign them to bundles.
6. You can edit old categories and you cannot change their type (assign it to different bundles). That would cause chaos when the category has some items in it. So I rather don't allow to change it.
7. Try to do crazy things with categories like people do.

### CODE REVIEW

**Comment by alanhartless** [comment]
@escopecz Does this support if a plugin wants to add multiple Category types (multiple per bundle)?

**Comment by escopecz** [comment]
Only issue I know about and don't know how to solve is pagination when the bundle (type) is not selected. It creates this URL to the second page `/s/categories/2`. As the default bundle the "category" bundle is used (https://github.com/escopecz/mautic/blob/category-on-one-place/app/bundles/CategoryBundle/Config/config.php#L17) so I don't know why it switch $bundle with $page and causes
```
app/bundles/CategoryBundle/Controller/CategoryController.php (line #58): at Mautic\CoreBundle\Security\Permissions\CorePermissions -> isGranted ( array('2:categories:view', '2:categories:create', '2:categories:edit', '2:categories:delete'), 'RETURN_ARRAY' )
```
It should be `category:categories:create` and so on. @alanhartless do you have an idea what's wrong? Pagination works when a bundle is filtered.

**Comment by escopecz** [comment]
@alanhartless yes, any bundle can subscribe like EmailBundle for example https://github.com/mautic/mautic/pull/864/files#diff-e84f92f6061c071ab522e17e065daf1fR1

**Comment by dbhurley** [comment]
+1, tested, and after clearing cache, the left column looks correct and has no submenus except for when appropriate. The configuration menu has a category menu item, and creating a category works and shows on the appropriate type.

**Comment by alanhartless** [comment]
Right, I can see any bundle can subscribe but what if my bundle/plugin wants two category groups? For example, if we wanted to categorize point actions and point triggers in two different category groups?  Can I do that? Or are we going to say one category per bundle? 

I'll have to play with the pagination.  My first thought was the baseUrl but it looks like that is set correctly just glancing at the code.

**Comment by escopecz** [comment]
I didn't test such a scenario. The only problem I can see is that default permissions has to exist for each category type. Other than that, it should work.

File: app/bundles/CategoryBundle/Event/CategoryBundlesEvent.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        asort($this->bundles);
+        return $this->bundles;
+    }
+
+    /**
+     * Adds the bundle entity
+     *
+     * @param string $bundle
+     */
+    public function addBundle($bundle)
```
Comment (alanhartless) [inline app/bundles/CategoryBundle/Event/CategoryBundlesEvent.php]: Could that problem be solved here if we allowed a second argument to take the place of the second $bundle and default to $bundle if null or empty?

Comment (escopecz) [inline app/bundles/CategoryBundle/Event/CategoryBundlesEvent.php]: What problem do you have in mind?

Comment (alanhartless) [inline app/bundles/CategoryBundle/Event/CategoryBundlesEvent.php]: Referring to your previous comment.

Comment (escopecz) [inline app/bundles/CategoryBundle/Event/CategoryBundlesEvent.php]: Is there any problem to call the `addBundle` method in one subscriber multiple times? For example in `PointBunlde\EventListener\CategorySubscriber` could be
```
public function onCategoryBundleListBuild(CategoryBundlesEvent $event)
{
    $event->addBundle('point-trigger');
    $event->addBundle('point-action');
}
```

**Comment by alanhartless** [comment]
I tested this with a plugin and it seems to work ok.  The only caveat is that a plugin requires the `plugin:` prefix when registering it as the bundle in order for the permission class to be found.  So it requires something like

```
    public function onCategoryBundleListBuild(CategoryBundlesEvent $event)
    {
        $event->addBundle('plugin:acme');
    }
```

That results in a language string of `mautic.plugin:acme.plugin:acme` which is not very pretty.

Can we change addBundle to accept an optional label instead of forcing `'mautic.' . $bundle . '.' . $bundle;` ? 

```
/**
     * Adds the bundle entity
     *
     * @param string $bundle
     */
    public function addBundle($bundle, $label = null)
    {
        if ($label == null) {
           $label = 'mautic.' . $bundle . '.' . $bundle;
        }
        $this->bundles[$bundle] = $label;
    }
```

There is a bug when editing a category, it changes the type/bundle value to "category" rather than preserving the original.  For example if I edit a "form" category and save, it'll become a "category" category.  

Just a few other little things to get thoughts on:

1. This is just semantics, but thoughts on changing "addBundle" to "addCategory" or "addCategoryType"  or the like since a bundle can technically have multiple categories? (BTW,  multiple categories do work as you stated John as long as the permission classes match up).  I know we refer to it as $bundle in our code elsewhere but it may help prevent confusion for 3rd party devs that won't necessarily be looking at our controller and view code. 
2. For the type column in the table, can we use the label there rather than the raw getBundle() value (we should be able to pass in the available list to obtain the label since we are already populating the filter with them).
3. Finally, when creating a new category, should the form's Type select default what's chosen in the list filter?

Thanks!

**Comment by escopecz** [comment]
Thanks for recommendations and testing Alan! Code is modified accordingly.

**Comment by escopecz** [comment]
There is still the problem with pagination I described above. Anyone have an idea how to fix it?

**Comment by escopecz** [comment]
There is a little hack to force pagination work. If we don't find a better solution, this PR works and can be tested and merged.

**Comment by alanhartless** [comment]
Another thing I just remembered.  Inline form "create a new category" options should default the type to the context of the form and just hide the Type select.

For example, if I'm adding a new Campaign and select "Create a new category" from the category dropdown, the category modal should not give an option to select Type rather have an hidden input with a value of "campaign"  I think the value is already passed to the form as $options['bundle'] so I think we can just check for that https://github.com/mautic/mautic/blob/af61c74ed66900f504b6520aa3870984b858ceb4/app/bundles/CategoryBundle/Form/Type/CategoryType.php#L48-48 as well.

File: app/bundles/CategoryBundle/Controller/CategoryController.php [hunk 2]
```diff
@@ -155,6 +183,7 @@ public function newAction ($bundle)
         $cancelled = $valid = false;
         $method    = $this->request->getMethod();
         $inForm    = ($method == 'POST') ? $this->request->request->get('category_form[inForm]', 0, true) : $this->request->get('inForm', 0);
+        $bundlesForm = $this->get('form.factory')->create('category_bundles_form');
```
Comment (alanhartless) [inline app/bundles/CategoryBundle/Controller/CategoryController.php]: Is this used anywhere?

**Comment by alanhartless** [comment]
As for the pagination hack, I think that works fine. The only other thing I can think of is maybe something like this in indexAction

```
if (is_numeric($bundle)) {
    $page    = $bundle;
    $bundle = 'category';
}
```

File: app/bundles/CategoryBundle/Controller/CategoryController.php (hunk 2 above)
Comment (escopecz) [inline app/bundles/CategoryBundle/Controller/CategoryController.php]: Doesn't seem to be. Removed.

**Comment by escopecz** [comment]
Good idea, Alan! Implemented.

**Comment by alanhartless** [comment]
Looks like the inline-form is fixed but the change now forces a hidden input with value="category" when creating a new category through the categories list page.   I'm assuming that wasn't intentional?

**Comment by escopecz** [comment]
@alanhartless categories can be defined in bundle config file as we discussed. Let me know if this is how you pictured it.

**Comment by alanhartless** [comment]
Exactly what I was thinking. My only thing (don't emoji hit me), let's use `categories` as the config key since more than one category can be defined and the other keys are plural as well.  Sorry! OCD kicking in there.

**Comment by escopecz** [comment]
Haha. No problem. I was wondering if singular or plural. I looked at menu that is singular and used singular as well. Should have look at more places :)

**Comment by dbhurley** [comment]
+1, @alanhartless anything else required here?

**Comment by alanhartless** [comment]
@dbhurley I don't think so. I believe it's good to go.

**Pull request merged**